text/0000-multi-part-examples-in-rustdoc/index.md

```diff
@@ -0,0 +1,231 @@
+
+Currently, there are three means of achieving this:
+
+1. Use rust comments inside the long example.
+  Downsides:
+    1. No markdown
+    1. If a comment overflows the rendered code block horizontally, scrolling is possible, yet cumbersome, especially when the comment is far from the bottom, where the scroll bar happens to be.
+1. Split long examples into multiple code blocks and add the text between them.
+  Downsides:
+    1. Must disable doc-test-ing the example
+    1. Lack of UI indicating that the code blocks are multiple parts of a whole
+1. [The current documented workaround][current-documented-workaround]
+  Downsides:
+    1. Tedious. Macros may make this more reasonable, but still not nice.
+
+This RFC suggests a new feature in rustdoc that supports multi-part examples without any of the downsides listed above.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## Multi-part examples
+
+A long example may benefit from being split up into multiple code blocks. The rendered documentation will indicate that each code block is a part of a whole. Also, doc-tests, run links, etc. will work as expected.
+
+The first block of a multi-part example must be tagged with `split-start`.
+Zero or more continuation blocks tagged with `split-continue` may follow.
+A final `split-end` block must be present.
+Between parts, any markdown may appear, including code blocks.
+However, no interleaving or nesting of multi-part examples is allowed.
+
+Not only _rust_ examples may be multi-part.
+
+Example:
+
+    //! ```split-start
+    //! let mut a = 5;
+    //! ```
+    //! 
+    //! We have declared a _mutable_ variable `a` of type `i32`.
+    //! 
+    //! ```split-continue
+    //! a = 10;
+    //! ```
+    //!
+    //! We have reassigned `a`. Now, let's look at the value:
+    //!
+    //! ```split-end
+    //! dbg!(a);
+    //! ```
+
+Doc-test related tags, such as `ignore` and `no-run` must be specified on the `split-start` block. Such tags specified on `split-continue` or `split-end` blocks will be ignored.
+
+Language is not automatically inherited from the one in the `split-start` block. If any part has a language tag that is different than the one in the `split-start`, the documentation will fail to build.
+
+If `split-start` / `split-continue` / `split-end` blocks appear in an erroneous order, the documentation will fail to build.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+> This is the technical portion of the RFC. Explain the design in sufficient detail that:
+>
+> - Its interaction with other features is clear.
+
+- Interaction with doc-test related tags seems to be fully explained in the previous section.
+- Interaction with langauge tags seems to be explained, other than the specific error details.
+
+> - It is reasonably clear how the feature would be implemented.
+
+We are not familiar with the code base.
+
+TODO:
+- UI indication mockup
```

Comment:
I would like to see this before going much further with the RFC, I think this is a big part of the feature that hasn't been addressed almost at all.

Reply:
Will do.

Reply:
Added in b920c62961ced755ca2c4a846bdb240eaf942e30.